source/expressions.tex

```diff
@@ -2883,7 +2883,7 @@
 constitute the arguments to the function. The postfix expression shall
 have function type or function pointer type.
 For a call to a non-member function or to a static member function,
-the postfix expression shall be either an lvalue that refers to a
+either the postfix expression shall be an lvalue that refers to a
 function (in which case the function-to-pointer standard
 conversion\iref{conv.func} is suppressed on the postfix expression),
 or it shall have function pointer type.
```

Comment:
This change is wrong. The "it" in the latter case is the postfix-expression, not the call. Suggest changing "shall be either" to "shall either be".

Reply:
Old code:

> For a call to a non-member function or to a static member function, the postfix expression shall be either an lvalue that refers to a function (in which case the function-to-pointer standard conversion [conv.func] is suppressed on the postfix expression), or it shall have function pointer type.

My version:

> For a call to a non-member function or to a static member function, either the postfix expression shall be an lvalue that refers to a function (in which case the function-to-pointer standard conversion [conv.func] is suppressed on the postfix expression), or it shall have function pointer type.

Your suggested version (IIUC):

> For a call to a non-member function or to a static member function, the postfix expression shall either be an lvalue that refers to a function (in which case the function-to-pointer standard conversion [conv.func] is suppressed on the postfix expression), or have function pointer type.


I think my version and your version have the same normative meaning, so I'm okay switching to your version, but I'd like to understand better why you interpret them as meaning different things.

Perhaps related: I encountered several places — this being one of them — where the normative text says "X either shall be Y or shall be Z." I read this as _literally_ saying that "it is required either that X be Y or that X be Z — but we're not going to tell you which of these two possible requirements is the actual requirement!"  Really, every single one of these instances ought to be recast as "X shall be either Y or Z." But sometimes that leads to harder-to-parse sentences, so generally I didn't do it in this patch.

Reply:
In your version, the "it" has a somewhat-ambiguous antecedent. A reader could parse it as either:

> For a call to a non-member function or to a static member function, either the postfix expression shall be an lvalue that refers to a function (in which case the function-to-pointer standard conversion [conv.func] is suppressed on the postfix expression), or [the postfix expression] shall have function pointer type.

or as

> For a call to a non-member function or to a static member function, either the postfix expression shall be an lvalue that refers to a function (in which case the function-to-pointer standard conversion [conv.func] is suppressed on the postfix expression), or [the call to a non-member function or to a static member function] shall have function pointer type.

depending on whether they think the scope of the "postfix expression" ends at the end of the clause of the "either" in which it was introduced. (Imagine that you skip from the "either" to the "or" when reading because you know you don't care about that clause.) My alternative avoids that problem.